src/lib/hooks/swap/useSwapInfo.tsx

```diff
@@ -118,5 +97,37 @@ export function SwapInfoUpdater() {
 
 /** Requires that SwapInfoUpdater be installed in the DOM tree. **/
 export default function useSwapInfo(): SwapInfo {
-  return useAtomValue(swapInfoAtom)
+  const swapInfo = useAtomValue(swapInfoAtom)
+
+  const { [Field.INPUT]: currencyIn, [Field.OUTPUT]: currencyOut } = useAtomValue(swapAtom)
+  const tradeState = useMemo(() => {
+    const { trade } = swapInfo
+    if (trade.state === TradeState.VALID && trade.trade) {
+      const isTradeStale =
+        (currencyIn && !trade.trade.inputAmount.currency.equals(currencyIn)) ||
+        (currencyOut && !trade.trade.outputAmount.currency.equals(currencyOut))
+      // swapInfo has not yet caught up to swapAtom.
+      if (isTradeStale) return TradeState.LOADING
+    }
+    return trade.state
+  }, [currencyIn, currencyOut, swapInfo])
+
+  const { account } = useActiveWeb3React()
+  const [balanceIn, balanceOut] = useCurrencyBalances(
```

Comment:
why do we specifically pull balance out of the other hook?

Reply:
I wanted to ensure that balance was being computed from the correct currency, and not from the stale currency.